### PR TITLE
feat: synchronize restricted courses from discovery

### PR DESCRIPTION
**Pull request by iloveagent57**
https://2u-internal.atlassian.net/browse/ENT-9570
This is one of two PRs required for the ticket. In this one, we'll deal only with synchronizing top-level restricted courses.

## Post-review

* [ ] Squash commits into discrete sets of changes
* [ ] Ensure that once the changes have been [deployed to stage](https://gocd.tools.edx.org/go/pipeline/activity/stage-enterprise_catalog), prod is manually deployed

### CODE REVIEW

File: enterprise_catalog/apps/catalog/models.py [hunk 1]
```diff
@@ -1307,30 +1361,65 @@ def update_contentmetadata_from_discovery(catalog_query, dry_run=False):
...
+    """
+    Fetch and assoicate any permitted restricted couress for the given catalog_query.
+    """
+    if not getattr(settings, 'SHOULD_FETCH_RESTRICTED_COURSE_RUNS', False):
+        return []
+
+    if not catalog_query.restricted_runs_allowed:
+        return []
+
+    restricted_course_keys = list(catalog_query.restricted_runs_allowed)
```
Comment (pwnage101) [inline enterprise_catalog/apps/catalog/models.py]: Nit: if the goal is to get the keys of a dict, I think dict.keys() is more explicit than list(dict).

File: enterprise_catalog/apps/catalog/models.py [hunk 2]
```diff
@@ -818,7 +821,58 @@ def __str__(self):
...
+        parent_record = ContentMetadata.objects.get(content_key=course_key)
+        record, _ = cls.objects.update_or_create(
+            content_key=course_key,
+            content_uuid=course_metadata_dict['uuid'],
+            content_type=COURSE,
+            unrestricted_parent=parent_record,
+            catalog_query=catalog_query,
+            defaults={
+                '_json_metadata': course_metadata_dict,
+            },
```
Comment (pwnage101) [inline enterprise_catalog/apps/catalog/models.py:853]: shouldn't we just overwrite _json_metadata instead of deferring to the existing one?

**Review by pwnage101 — COMMENTED** [review]
Looks good so far!

File: enterprise_catalog/apps/catalog/models.py (hunk 2 above)
Comment (iloveagent57) [reply enterprise_catalog/apps/catalog/models.py:853]: This is what `defaults` does for `update_or_create` (and why"defaults" is a poorly-named keyward arg):
> The update_or_create method tries to fetch an object from database based on the given kwargs. If a match is found, it updates the fields passed in the defaults dictionary.
https://docs.djangoproject.com/en/4.2/ref/models/querysets/#update-or-create

Comment (pwnage101) [reply enterprise_catalog/apps/catalog/models.py:853]: :shriek: I fear I've been misusing update_or_create for years now...

**Review by pwnage101 — APPROVED** [review]
LGTM!

**Pull request merged**